From 8fda9d4d89ffd98883717170f67beec2bba0f777 Mon Sep 17 00:00:00 2001
From: huangyifeng <huangyifeng@eswincomputing.com>
Date: Wed, 24 Jul 2024 15:10:18 +0800
Subject: [PATCH 117/533] fix(irqchip/sifive-plic): enable interrupt if needed
 before EOI

Changelogs:

there is a small window after a hart
claiming an interrupt and before irq_desc->lock getting acquired,
interrupt can be disabled during this window. Thus, EOI can be invoked
while the interrupt is disabled, effectively nullify this EOI.
Make sure that interrupt is really enabled before EOI
(Merge from linux 6.10)

Signed-off-by: huangyifeng <huangyifeng@eswincomputing.com>
---
 drivers/irqchip/irq-sifive-plic.c | 10 ++++++++--
 1 file changed, 8 insertions(+), 2 deletions(-)

diff --git a/drivers/irqchip/irq-sifive-plic.c b/drivers/irqchip/irq-sifive-plic.c
index a8f5cfad16f7..a0ee67cb06b1 100644
--- a/drivers/irqchip/irq-sifive-plic.c
+++ b/drivers/irqchip/irq-sifive-plic.c
@@ -136,6 +136,10 @@ static void plic_irq_mask(struct irq_data *d)
 
 static void plic_irq_enable(struct irq_data *d)
 {
+#ifdef CONFIG_PM_SLEEP
+	struct irq_desc *desc = irq_data_to_desc(d);
+	/* Avoid reporting errors when removing irq. */
+#endif
 	plic_irq_toggle(irq_data_get_effective_affinity_mask(d), d, 1);
 	plic_irq_unmask(d);
 }
@@ -149,7 +153,7 @@ static void plic_irq_eoi(struct irq_data *d)
 {
 	struct plic_handler *handler = this_cpu_ptr(&plic_handlers);
 
-	if (unlikely(irqd_irq_disabled(d))) {
+	if (irqd_irq_disabled(d)) {
 		plic_toggle(handler, d->hwirq, 1);
 		writel(d->hwirq, handler->hart_base + CONTEXT_CLAIM);
 		plic_toggle(handler, d->hwirq, 0);
@@ -165,6 +169,7 @@ static int plic_set_affinity(struct irq_data *d,
 	unsigned int cpu;
 	struct cpumask amask;
 	struct plic_priv *priv = irq_data_get_irq_chip_data(d);
+	struct irq_desc *desc = irq_data_to_desc(d);
 
 	cpumask_and(&amask, &priv->lmask, mask_val);
 
@@ -173,8 +178,9 @@ static int plic_set_affinity(struct irq_data *d,
 	else
 		cpu = cpumask_any_and(&amask, cpu_online_mask);
 
-	if (cpu >= nr_cpu_ids)
+	if (cpu >= nr_cpu_ids) {
 		return -EINVAL;
+	}
 
 	/* Invalidate the original routing entry */
 	plic_irq_toggle(irq_data_get_effective_affinity_mask(d), d, 0);
-- 
2.53.0

